File: core/src/test/java/uk/lobsterdoodle/namepicker/selection/SelectionNamesUseCaseTest.java
package uk.lobsterdoodle.namepicker.selection;

import org.junit.Before;
import org.junit.Test;

import java.util.Collections;

import uk.lobsterdoodle.namepicker.events.EventBus;
import uk.lobsterdoodle.namepicker.model.Name;
import uk.lobsterdoodle.namepicker.storage.GroupNamesRetrievedEvent;
import uk.lobsterdoodle.namepicker.storage.KeyValueStore;

import static java.util.Arrays.asList;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static uk.lobsterdoodle.namepicker.selection.SelectionNamesUseCase.GROUP_SORT_TYPE_FOR_GROUP_FORMAT;

public class SelectionNamesUseCaseTest {
    private SelectionNamesUseCase useCase;
    private EventBus bus;
    private KeyValueStore store;
    private KeyValueStore.Edit edit;

    @Before
    public void setUp() {
        bus = mock(EventBus.class);
        store = mock(KeyValueStore.class);
        edit = mock(KeyValueStore.Edit.class);
        when(store.edit()).thenReturn(edit);
        when(edit.put(anyString(), anyString())).thenReturn(edit);
        useCase = new SelectionNamesUseCase(bus, store);
    }

    @Test
    public void registers_on_bus() {
        verify(bus).register(useCase);
    }

    @Test
    public void on_GroupNamesRetrievedEvent_when_sort_type_is_none_post_GroupNamesSortedEvent() {
        when(store.getString(String.format(GROUP_SORT_TYPE_FOR_GROUP_FORMAT, 24L), NameSortType.NONE.toString())).thenReturn(NameSortType.NONE.toString());
        useCase.on(new GroupNamesRetrievedEvent(24L, asList(name(2L, "kim"), name(1L, "Bauer"), name(3L, "Yelena"))));
        verify(bus).post(new GroupNamesSortedEvent(asList(name(2L, "kim"), name(1L, "Bauer"), name(3L, "Yelena"))));
    }

    @Test
    public void on_GroupNamesRetrievedEvent_when_sort_type_is_asc_post_GroupNamesSortedEvent() {
        when(store.getString(String.format(GROUP_SORT_TYPE_FOR_GROUP_FORMAT, 24L), NameSortType.NONE.toString())).thenReturn(NameSortType.ASC.toString());
        useCase.on(new GroupNamesRetrievedEvent(24L, asList(name(2L, "kim"), name(1L, "Bauer"), name(3L, "Yelena"))));
        verify(bus).post(new GroupNamesSortedEvent(asList(name(1L, "Bauer"), name(2L, "kim"), name(3L, "Yelena"))));
    }

    @Test
    public void on_GroupNamesRetrievedEvent_when_sort_type_is_desc_post_GroupNamesSortedEvent() {
        when(store.getString(String.format(GROUP_SORT_TYPE_FOR_GROUP_FORMAT, 24L), NameSortType.NONE.toString())).thenReturn(NameSortType.DESC.toString());
        useCase.on(new GroupNamesRetrievedEvent(24L, asList(name(2L, "kim"), name(1L, "Bauer"), name(3L, "Yelena"))));
        verify(bus).post(new GroupNamesSortedEvent(asList(name(3L, "Yelena"), name(2L, "kim"), name(1L, "Bauer"))));
    }

    @Test
    public void on_SortMenuItemSelectedEvent_when_sort_type_is_none_write_new_sort_type_asc_to_storage() {
        when(store.getString(String.format(GROUP_SORT_TYPE_FOR_GROUP_FORMAT, 24L), NameSortType.NONE.toString())).thenReturn(NameSortType.NONE.toString());
        useCase.on(new SortMenuItemSelectedEvent(24L, Collections.emptyList()));
        verify(edit).put("group_sort_type_for_group_24", NameSortType.ASC.toString());
        verify(edit).commit();
    }

    @Test
    public void on_GroupNamesRetrievedEvent_when_sort_type_is_asc_write_new_sort_type_desc_to_storage() {
        when(store.getString(String.format(GROUP_SORT_TYPE_FOR_GROUP_FORMAT, 24L), NameSortType.NONE.toString())).thenReturn(NameSortType.ASC.toString());
        useCase.on(new SortMenuItemSelectedEvent(24L, Collections.emptyList()));
        verify(edit).put("group_sort_type_for_group_24", NameSortType.DESC.toString());
        verify(edit).commit();
    }

    @Test
    public void on_GroupNamesRetrievedEvent_when_sort_type_is_desc_write_new_sort_type_asc_to_storage() {
        when(store.getString(String.format(GROUP_SORT_TYPE_FOR_GROUP_FORMAT, 24L), NameSortType.NONE.toString())).thenReturn(NameSortType.DESC.toString());
        useCase.on(new SortMenuItemSelectedEvent(24L, Collections.emptyList()));
        verify(edit).put("group_sort_type_for_group_24", NameSortType.ASC.toString());
        verify(edit).commit();
    }

    @Test
    public void on_SortMenuItemSelectedEvent_post_GroupNamesSortedEvent() {
        when(store.getString(String.format(GROUP_SORT_TYPE_FOR_GROUP_FORMAT, 24L), NameSortType.NONE.toString())).thenReturn(NameSortType.NONE.toString());
        useCase.on(new SortMenuItemSelectedEvent(24L, asList(name(2L, "kim"), name(1L, "Bauer"), name(3L, "Yelena"))));
        verify(bus).post(new GroupNamesSortedEvent(asList(name(1L, "Bauer"), name(2L, "kim"), name(3L, "Yelena"))));
    }

    private Name name(long id, String name) {
        return new Name(id, name, false);
    }
}

File: core/src/test/java/uk/lobsterdoodle/namepicker/selection/SelectionNamesUseCaseTest.kt
package uk.lobsterdoodle.namepicker.selection

import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.whenever
import org.junit.Before
import org.junit.Test
import uk.lobsterdoodle.namepicker.Testing.Util.anyString

import uk.lobsterdoodle.namepicker.events.EventBus
import uk.lobsterdoodle.namepicker.model.Name
import uk.lobsterdoodle.namepicker.storage.GroupNamesRetrievedEvent
import uk.lobsterdoodle.namepicker.storage.KeyValueStore

class SelectionNamesUseCaseTest {
    private lateinit var useCase: SelectionNamesUseCase
    private lateinit var bus: EventBus
    private lateinit var store: KeyValueStore
    private lateinit var edit: KeyValueStore.Edit

    @Before
    fun setUp() {
        bus = mock()
        store = mock()
        edit = mock()
        whenever(store.edit()).thenReturn(edit)
        whenever(edit.put(anyString(), anyString())).thenReturn(edit)
        useCase = SelectionNamesUseCase(bus, store)
    }

    @Test
    fun registers_on_bus() {
        verify(bus).register(useCase)
    }

    @Test
    fun on_GroupNamesRetrievedEvent_when_sort_type_is_none_post_GroupNamesSortedEvent() {
        whenever(store.getString("group_sort_type_for_group_24", NameSortType.NONE.toString())).thenReturn(NameSortType.NONE.toString())
        useCase.on(GroupNamesRetrievedEvent(24L, listOf(name(2L, "kim"), name(1L, "Bauer"), name(3L, "Yelena"))))
        verify(bus).post(GroupNamesSortedEvent(listOf(name(2L, "kim"), name(1L, "Bauer"), name(3L, "Yelena"))))
    }

    @Test
    fun on_GroupNamesRetrievedEvent_when_sort_type_is_asc_post_GroupNamesSortedEvent() {
        whenever(store.getString("group_sort_type_for_group_24", NameSortType.NONE.toString())).thenReturn(NameSortType.ASC.toString())
        useCase.on(GroupNamesRetrievedEvent(24L, listOf(name(2L, "kim"), name(1L, "Bauer"), name(3L, "Yelena"))))
        verify(bus).post(GroupNamesSortedEvent(listOf(name(1L, "Bauer"), name(2L, "kim"), name(3L, "Yelena"))))
    }

    @Test
    fun on_GroupNamesRetrievedEvent_when_sort_type_is_desc_post_GroupNamesSortedEvent() {
        whenever(store.getString("group_sort_type_for_group_24", NameSortType.NONE.toString())).thenReturn(NameSortType.DESC.toString())
        useCase.on(GroupNamesRetrievedEvent(24L, listOf(name(2L, "kim"), name(1L, "Bauer"), name(3L, "Yelena"))))
        verify(bus).post(GroupNamesSortedEvent(listOf(name(3L, "Yelena"), name(2L, "kim"), name(1L, "Bauer"))))
    }

    @Test
    fun on_SortMenuItemSelectedEvent_when_sort_type_is_none_write_new_sort_type_asc_to_storage() {
        whenever(store.getString("group_sort_type_for_group_24", NameSortType.NONE.toString())).thenReturn(NameSortType.NONE.toString())
        useCase.on(SortMenuItemSelectedEvent(24L, emptyList<Name>()))
        verify(edit).put("group_sort_type_for_group_24", NameSortType.ASC.toString())
        verify(edit).commit()
    }

    @Test
    fun on_GroupNamesRetrievedEvent_when_sort_type_is_asc_write_new_sort_type_desc_to_storage() {
        whenever(store.getString("group_sort_type_for_group_24", NameSortType.NONE.toString())).thenReturn(NameSortType.ASC.toString())
        useCase.on(SortMenuItemSelectedEvent(24L, emptyList<Name>()))
        verify(edit).put("group_sort_type_for_group_24", NameSortType.DESC.toString())
        verify(edit).commit()
    }

    @Test
    fun on_GroupNamesRetrievedEvent_when_sort_type_is_desc_write_new_sort_type_asc_to_storage() {
        whenever(store.getString("group_sort_type_for_group_24", NameSortType.NONE.toString())).thenReturn(NameSortType.DESC.toString())
        useCase.on(SortMenuItemSelectedEvent(24L, emptyList<Name>()))
        verify(edit).put("group_sort_type_for_group_24", NameSortType.ASC.toString())
        verify(edit).commit()
    }

    @Test
    fun on_SortMenuItemSelectedEvent_post_GroupNamesSortedEvent() {
        whenever(store.getString("group_sort_type_for_group_24", NameSortType.NONE.toString())).thenReturn(NameSortType.NONE.toString())
        useCase.on(SortMenuItemSelectedEvent(24L, listOf(name(2L, "kim"), name(1L, "Bauer"), name(3L, "Yelena"))))
        verify(bus).post(GroupNamesSortedEvent(listOf(name(1L, "Bauer"), name(2L, "kim"), name(3L, "Yelena"))))
    }

    private fun name(id: Long, name: String): Name = Name(id, name, false)
}